packages/wrangler/src/proxy.ts

```diff
@@ -101,3 +36,264 @@ const HTTP1_HEADERS = new Set([
+          responseHeaders,
+          previewToken.host,
+          port
+        );
+        stream.respond(responseHeaders);
+        request.pipe(stream, { end: true });
+      });
+    }
+    proxy.on("stream", handleStream);
+    cleanupListeners.push(() => proxy.off("stream", handleStream));
+
+    // flush and replay buffered streams
+    streamBufferRef.current.forEach((buffer) =>
+      handleStream(buffer.stream, buffer.headers)
+    );
+    streamBufferRef.current = [];
+
+    /** HTTP/1 -> HTTP/2  */
+    const handleRequest: RequestListener = (
+      message: IncomingMessage,
+      response: ServerResponse
+    ) => {
+      const { httpVersionMajor, headers, method, url } = message;
+      if (httpVersionMajor >= 2) {
+        return; // Already handled by the "stream" event.
+      }
+      addCfPreviewTokenHeader(headers, previewToken.value);
+      headers[":method"] = method;
+      headers[":path"] = url;
+      headers[":authority"] = previewToken.host;
+      headers[":scheme"] = "https";
+      for (const name of Object.keys(headers)) {
+        if (HTTP1_HEADERS.has(name.toLowerCase())) {
+          delete headers[name];
+        }
+      }
+      const request = message.pipe(remote.request(headers));
+      request.on("response", (responseHeaders) => {
+        const status = responseHeaders[":status"];
+        rewriteRemoteHostToLocalHostInHeaders(
+          responseHeaders,
+          previewToken.host,
+          port
+        );
+        for (const name of Object.keys(responseHeaders)) {
+          if (name.startsWith(":")) {
+            delete responseHeaders[name];
+          }
+        }
+        response.writeHead(status, responseHeaders);
+        request.pipe(response, { end: true });
+      });
+    };
+    // If an asset path is defined, check the file system
+    // for a file first and serve if it exists.
+    if (assetPath) {
+      const handleAsset = serveStatic(assetPath, {
+        cacheControl: false,
+      });
+      const handleAssetsAndRequests = (
+        request: IncomingMessage,
+        response: ServerResponse
+      ) => {
+        handleAsset(request, response, () => {
+          handleRequest(request, response);
+        });
+      };
+
+      proxy.on("request", handleAssetsAndRequests);
+      cleanupListeners.push(() =>
+        proxy.off("request", handleAssetsAndRequests)
+      );
+
+      // flush and replay buffered requests
+      requestResponseBufferRef.current.forEach(({ request, response }) =>
+        handleAssetsAndRequests(request, response)
+      );
+      requestResponseBufferRef.current = [];
+    } else {
+      proxy.on("request", handleRequest);
+      cleanupListeners.push(() => proxy.off("request", handleRequest));
+      // flush and replay buffered requests
+      requestResponseBufferRef.current.forEach(({ request, response }) =>
+        handleRequest(request, response)
+      );
+      requestResponseBufferRef.current = [];
+    }
```

Comment:
NIT - this whole block could be simplified to remove some of the duplication - for example:

```ts
const actualHandleRequest = assetPath ?
    createHandleAssetsRequest(assetPath, handleRequest) :
    handleRequest;

proxy.on("request", actualHandleRequest);
cleanupListeners.push(() => proxy.off("request", actualHandleRequest));

// flush and replay buffered requests
requestResponseBufferRef.current.forEach(({ request, response }) =>
  actualHandleRequest(request, response)
);
requestResponseBufferRef.current = [];

// ... somewhere else

function createHandleAssetsRequest(assetPath: string, handleRequest: RequestListener) {
  const handleAsset = serveStatic(assetPath, {
    cacheControl: false,
  });
  return (
    request: IncomingMessage,
    response: ServerResponse
  ) => {
    handleAsset(request, response, () => {
      handleRequest(request, response);
    };
  };
}
```

Reply:
I like this!